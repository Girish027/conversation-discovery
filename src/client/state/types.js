// Add types that will be used across various reducers here.

export const LOCATION_CHANGE = '@@router/LOCATION_CHANGE';
export const UPDATE_CAA = 'all/UPDATE_CAA';
export const SELECT_ACTIVE_PROJECT = 'projects/SELECT_ACTIVE_PROJECT';
export const SELECT_RUN = 'runs/SELECT_RUN';
export const SELECT_CLUSTER = 'clusters/SELECT_CLUSTER';
export const DOWNLOAD_FILE = 'all/DOWNLOAD_FILE';
export const SET_PROJECT_COUNT = 'projects/SET_PROJECT_COUNT';
export const SET_RUN_COUNT = 'runs/SET_RUN_COUNT';

// Action for testing
export const NOOP = 'NOOP';
