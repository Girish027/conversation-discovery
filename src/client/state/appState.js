import {
  fromJS, List, Map
} from 'immutable';
import { createSelector } from 'reselect';
import axios from 'axios';
import _ from 'lodash';
import { push } from 'connected-react-router';
import { getURLParams } from 'utils/URIUtils';
import { routeNames, getRoute } from 'utils/RouteHelper';
import { getPathnameSelector } from 'state/routerState';
import appLogger from 'utils/appLogger';
import { LOCATION_CHANGE, UPDATE_CAA } from 'state/types';
import { getAllProjects } from 'state/projectsState';
import { setDefaultsForAPI } from 'utils/api';
import { isUserAuthorized, getSortedClientsList } from 'utils/caaHelper';
import constants from '../constants';
import { clearStorage, loadStorage, saveStorage } from '../utils/storageManager';
import { logAmplitudeEvent } from '../utils/amplitudeUtils';
import { roles, updateRoleConfig } from '../utils/roleHelper';

export const types = {
  SET_USER_CONFIG: 'app/SET_USER_CONFIG',
  SET_CLIENTS_CONFIG: 'app/SET_CLIENTS_CONFIG',
  MODAL_IS_OPEN: 'app/MODAL_IS_OPEN',
  MODAL_MESSAGE: 'app/MODAL_MESSAGE',
  INIT_ANALYTICS: 'app/INIT_ANALYTICS',
  SET_UNIFIED_PORTAL_CLIENT_PICKER: 'app/SET_UNIFIED_PORTAL_CLIENT_PICKER',
  SHOW_WORD_CLOUD: 'app/SHOW_WORD_CLOUD',
};

export const stateKey = {
  authEnabled: 'authEnabled',
  clientId: 'clientId',
  componentClientId: 'componentClientId',
  appId: 'appId',
  accountId: 'accountId',
  userName: 'userName',
  userType: 'userType',
  itsBaseUrl: 'itsBaseUrl',
  analyticsKey: 'analyticsKey',
  clientConfigs: 'clientConfigs',
  modalIsOpen: 'modalIsOpen',
  modalMessage: 'modalMessage',
  modalState: 'modalState',
  unifiedPortalClientPicker: 'unifiedPortalClientPicker',
  docPortalUrl: 'docPortalUrl',
  unifiedPortalUrl: 'unifiedPortalUrl',
  contactSupportUrl: 'contactSupportUrl',
  oktaUserAccountUrl: 'oktaUserAccountUrl',
  isAnswers: 'isAnswers',
  isConversations: 'isConversations',
  userRole: 'userRole',
  roleConfig: 'roleConfig',
  showWordCloud: 'showWordCloud',
  wordCloudcontent: 'wordCloudcontent',
};

export const initialAppState = fromJS({
  authEnabled: false,
  clientId: '',
  componentClientId: '',
  appId: '',
  accountId: '',
  userName: '',
  userType: '',
  itsBaseUrl: '',
  analyticsKey: '',
  clientConfigs: List(),
  modalIsOpen: false,
  modalMessage: '',
  modalState: {},
  unifiedPortalClientPicker: false,
  docPortalUrl: '',
  unifiedPortalUrl: '',
  contactSupportUrl: '',
  oktaUserAccountUrl: '',
  isConversations: false,
  isAnswers: false,
  userRole: '',
  roleConfig: {},
  showWordCloud: false,
  wordCloudcontent: '',
});

const devlog = appLogger({ tag: 'appState' });

export const updateCAA = (caa) => ({
  type: UPDATE_CAA,
  caa,
});

export const setUnifiedPortalClientPicker = (unifiedPortalClientPicker) => ({
  type: types.SET_UNIFIED_PORTAL_CLIENT_PICKER,
  unifiedPortalClientPicker
});

export const setUserConfig = (config) => ({
  type: types.SET_USER_CONFIG,
  config,
});

export const setClientsConfig = (clientConfigs = []) => ({
  type: types.SET_CLIENTS_CONFIG,
  clientConfigs,
});

export const setShowWordCloud = (showWordCloud = false, wordCloudcontent = '') => ({
  type: types.SHOW_WORD_CLOUD,
  showWordCloud,
  wordCloudcontent,
});

export const setModalIsOpen = (modalIsOpen, modalState = {}) => ({
  type: types.MODAL_IS_OPEN,
  modalIsOpen,
  modalState,
});

export const setModalMessageState = (modalMessage) => ({
  type: types.MODAL_MESSAGE,
  modalMessage
});

export const setModalMessage = (modalMessage) => (dispatch) => {
  dispatch(setModalMessageState(modalMessage));
};

export const getConversationsConfig = () => (dispatch, getState) => {
  axios.get(constants.serverApiUrls.conversationsConfigUrl, { withCredentials: true })
    .then((configRes) => {
      const { data: configBody = {} } = configRes;
      const conversationsClientConfigs = configBody.clientConfigs;
      let clientConfigs = getClientConfigsSelector(getState());
      clientConfigs = clientConfigs.map((element) => {
        let configObj = element;
        const { clientId } = element;
        const conversationsClient = conversationsClientConfigs.find((config) => config.clientId === clientId);
        if (conversationsClient !== undefined) {
          configObj = { ...configObj, isConversations: true };
        }
        return configObj;
      });
      dispatch(setClientsConfig(clientConfigs));
    })
    .catch((configErr) => {
      devlog.error(configErr);
    });
};

export const getAnswersConfig = () => (dispatch, getState) => {
  axios.get(constants.serverApiUrls.answersConfigUrl, { withCredentials: true })
    .then((configRes) => {
      const { data: configBody = {} } = configRes;
      const answersClientConfigs = configBody.clientConfigs;
      let clientConfigs = getClientConfigsSelector(getState());
      clientConfigs = clientConfigs.map((element) => {
        let configObj = element;
        const { clientId } = element;
        const answersClient = answersClientConfigs.find((config) => config.clientId === clientId);
        if (answersClient !== undefined) {
          configObj = { ...configObj, isAnswers: true };
        }
        return configObj;
      });
      dispatch(setClientsConfig(clientConfigs));
    })
    .catch((configErr) => {
      devlog.error(configErr);
    });
};

export const getUserConfig = () => (dispatch, getState) => {
  axios.get(constants.serverApiUrls.userConfigUrl, { withCredentials: true })
    .then((configRes) => {
      const { data: configBody = {} } = configRes;
      setDefaultsForAPI();

      const sortClientConfigs = getSortedClientsList([...configBody.clientConfigs]);
      dispatch(setClientsConfig(sortClientConfigs));
      dispatch(getAnswersConfig());
      dispatch(getConversationsConfig());
      const userSelectedCAA = getCAASelector(getState());

      const { authorized, clientPicker, caa } = isUserAuthorized(userSelectedCAA, sortClientConfigs);
      if (clientPicker) {
        dispatch(setUnifiedPortalClientPicker(true));
      } else if (!authorized && (userSelectedCAA.clientId !== 'undefined' || userSelectedCAA.appId !== 'undefined')) {
        dispatch(setModalIsOpen(true, {
          modalName: constants.modals.unauthorized,
          header: 'Access Denied',
          message: `You are not authorized to view the client: ${userSelectedCAA.clientId}, app: ${userSelectedCAA.appId} and account: ${userSelectedCAA.accountId}`,
          onClickOk: () => dispatch(closeUnauthorizedModal()),
          onClickClose: () => dispatch(closeUnauthorizedModal()),
          onClickCancel: () => dispatch(closeUnauthorizedModal()),
        }));
      } else {
        dispatch(changeClient(caa));
      }
      dispatch(setUserConfig(configBody));
    })
    .catch((configErr) => {
      devlog.error(configErr);
    });
};

const closeUnauthorizedModal = () => (dispatch) => {
  dispatch(setUnifiedPortalClientPicker(true)); dispatch(setModalIsOpen(false));
};

// When user changes client
export const changeClient = (caa) => (dispatch, getState) => {
  const state = getState();
  const existingCAA = getCAASelector(state);
  if (!_.isEqual(existingCAA, caa)) {
    dispatch(updateCAA(caa));
    clearStorage(constants.activeProjectId);
    clearStorage(constants.isConversations);
    clearStorage(constants.isAnswers);
    clearStorage(constants.userRole);
  }
  const { clientId, appId } = caa;
  const obj = {
    clientId, appId, toolId: constants.toolName, env: constants.environment
  };
  logAmplitudeEvent('IDTClients', obj);
  dispatch(loadRoute(caa));
  dispatch(getAllProjects(caa));
};

const loadRoute = (caa) => (dispatch, getState) => {
  // refresh the page only if it is HOMEPAGE/baseRoute.
  const state = getState();
  const currentPath = getPathnameSelector(state);
  if (currentPath === `${routeNames.HOME_PAGE}/` || currentPath === `${routeNames.BASE_ROUTE}/` || currentPath === `${routeNames.DISCOVER_INTENTS}`) {
    const projectId = loadStorage(constants.activeProjectId, '');
    const newRoute = getRoute(routeNames.DISCOVER_INTENTS, { ...caa, projectId });
    dispatch(push(newRoute));
  }
};

// REDUCERS
export const appReducer = (state = initialAppState, action) => {
  switch (action.type) {
    case types.SET_CLIENTS_CONFIG:
      return state.set(stateKey.clientConfigs, List(action.clientConfigs));
    case types.SET_USER_CONFIG:
      return state.merge(Map({
        userName: action.config.username,
        userType: action.config.userType,
        authEnabled: action.config.authEnabled,
        itsBaseUrl: action.config.itsBaseUrl,
        analyticsKey: action.config.analyticsKey,
        docPortalUrl: action.config.docPortalUrl,
        unifiedPortalUrl: action.config.unifiedPortalUrl,
        contactSupportUrl: action.config.contactSupportUrl,
        oktaUserAccountUrl: action.config.oktaUserAccountUrl,
      }));
    case types.SET_UNIFIED_PORTAL_CLIENT_PICKER:
      return state.set(stateKey.unifiedPortalClientPicker, action.unifiedPortalClientPicker);
    case types.MODAL_IS_OPEN:
      return state.merge(Map({
        modalIsOpen: action.modalIsOpen,
        modalState: action.modalState,
      }));
    case types.MODAL_MESSAGE:
      return state.set(stateKey.modalMessage, action.modalMessage);
    case types.SHOW_WORD_CLOUD:
      return state.merge(Map({
        showWordCloud: action.showWordCloud,
        wordCloudcontent: action.wordCloudcontent,
      }));
    // The below reducers are handling part of the global state change
    case UPDATE_CAA:
      return state.merge(Map(action.caa));
    case LOCATION_CHANGE: {
      const { search } = action.payload.location;
      if (search) {
        const {
          clientid, appid, accountid, componentClientId
        } = getURLParams(search);

        let isAnswers = false;
        let isConversations = false;

        const client = state.get(stateKey.clientId);
        const clientConfigs = state.get(stateKey.clientConfigs);
        const selectedClientConfig = clientConfigs.find((config) => config.clientId === clientid);
        let userRole = roles.unauthorized;
        if (selectedClientConfig) {
          userRole = selectedClientConfig.role;
        }

        if (selectedClientConfig && selectedClientConfig.isAnswers) {
          isAnswers = true;
        }
        if (selectedClientConfig && selectedClientConfig.isConversations) {
          isConversations = true;
        }

        if (client === '') {
          isConversations = loadStorage(constants.isConversations, false);
          isAnswers = loadStorage(constants.isAnswers, false);
          userRole = loadStorage(constants.userRole, roles.unauthorized);
        } else {
          saveStorage(constants.isConversations, isConversations);
          saveStorage(constants.isAnswers, isAnswers);
          saveStorage(constants.userRole, userRole);
        }
        const roleConfig = updateRoleConfig(userRole);
        return state.merge(Map({
          clientId: clientid,
          appId: appid,
          accountId: accountid,
          componentClientId,
          isAnswers,
          isConversations,
          userRole,
          roleConfig,
        }));
      }
      return state;
    }
    default:
      return state;
  }
};

// SELECTORS
export const getAppState = (state) => state.get('app');

export const getCAASelector = createSelector(
  getAppState,
  (appState) => ({
    clientId: appState.get(stateKey.clientId),
    appId: appState.get(stateKey.appId),
    accountId: appState.get(stateKey.accountId),
    componentClientId: appState.get(stateKey.componentClientId),
  })
);

export const getClientIdSelector = createSelector(
  getCAASelector,
  (caa) => caa.clientId
);

export const getComponentClientIdSelector = createSelector(
  getCAASelector,
  (caa) => caa.componentClientId
);

export const getAppIdSelector = createSelector(
  getCAASelector,
  (caa) => caa.appId
);

export const getIsAnswersSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.isAnswers)
);

export const getIsConversationsSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.isConversations)
);

export const getAccountIdSelector = createSelector(
  getCAASelector,
  (caa) => caa.accountId
);

export const getClientConfigsSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.clientConfigs)
);

export const getITSBaseUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.itsBaseUrl)
);

export const getShowWordCloudSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.showWordCloud)
);

export const getWordCloudContentSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.wordCloudcontent)
);

export const getModalOpenSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.modalIsOpen)
);

export const getModalStateSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.modalState)
);

export const getModalMessageSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.modalMessage)
);

export const getUnifiedPortalClientPicker = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.unifiedPortalClientPicker)
);

export const getDocPortalUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.docPortalUrl)
);

export const getUnifiedPortalUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.unifiedPortalUrl)
);

export const getUserTypeSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.userType)
);

export const getContactSupportUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.contactSupportUrl)
);

export const getUserRoleSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.userRole)
);

export const getRoleConfigSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.roleConfig)
);

export const getOktaUserAccountUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.oktaUserAccountUrl)
);
