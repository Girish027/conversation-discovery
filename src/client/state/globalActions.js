import { downloadFile } from 'utils/FileUtils';
import { getCAASelector } from 'state/appState';
import { getActiveProjectId } from 'state/projectsState';
import Constants from '../constants';
import { logAmplitudeEvent } from '../utils/amplitudeUtils';
import {
  DOWNLOAD_FILE
} from './types';

export const downloadResults = (constructURL, runId, clusterId = '') => (dispatch, getState) => {
  const state = getState();
  const projectId = getActiveProjectId(state);
  const caa = getCAASelector(state);
  const locationUrl = constructURL({
    ...caa, projectId, runId, clusterId
  });
  dispatch({ type: DOWNLOAD_FILE, locationUrl });
  const downloadRun = { toolId: Constants.toolName, runId, env: Constants.environment };
  logAmplitudeEvent('IDT_DownloadRunResults', downloadRun);
  downloadFile(locationUrl);
};
