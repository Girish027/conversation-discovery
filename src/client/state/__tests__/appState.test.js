import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {
  fromJS, List
} from 'immutable';
import Constants from 'components/../constants';
import * as appState from 'state/appState';
import * as GlobalTypes from 'state/types';
import axios from 'axios';

const mockClientConfigData = [{
  "authEnabled":true,
  "itsBaseUrl":"https://stable.developer.sv2.247-inc.net/v1/integratedtoolsuite/",
  "analyticsKey":"",
  "clientConfigs":[
     {
        "clientId":"testclientone",
        "clientDisplayName":"Test Client One",
        "accounts":[
           {
              "accountId":"testclientone",
              "accountDisplayName":"Default",
              "packageCode":"custom",
              "products":[
                 {
                    "productId":"discovery",
                    "roles":[
                       "developer"
                    ],
                    "components":[
                       {
                          "componentId":"dialogmanager",
                          "componentClientId":"testclientone",
                          "componentAccountId":"testclientone"
                       }
                    ]
                 }
              ]
           }
        ],
        "componentClientId":"testclientone",
        "apps":[
           {
              "appId":"e2etesting",
              "accountId":"testclientone",
              "enableAiva":true,
              "enableAnswers7":false
           },
           {
              "appId":"general",
              "accountId":"testclientone",
              "enableAiva":true,
              "enableAnswers7":true
           },
           {
              "appId":"integration",
              "accountId":"testclientone",
              "enableAiva":true,
              "enableAnswers7":false
           },
           {
              "appId":"testclientone",
              "accountId":"testclientone",
              "enableAiva":true,
              "enableAnswers7":false
           }
        ]
     }
  ],
  "username":"Test user",
  "userType":"Internal",
  "clientId":"tfsai",
  "appId":"referencebot",
  "componentClientId":"247ai",
  "unifiedPortalUrl":"https://stable.developer.sv2.247-inc.net/home/",
  "docPortalUrl":"https://stable.developer.sv2.247-inc.net/docportal/Content/Intent-Discovery/Intent-Discovery.htm",
  "contactSupportUrl":"https://247inc.atlassian.net/secure/RapidBoard.jspa?rapidView=1178&projectKey=SRE"
}];


jest.mock('axios');
jest.mock('utils/caaHelper')

describe('state/App:', function () {
  const caa = {
    clientId: '247ai',
    appId: 'referencebot',
    accountId: 'referencebot'
  };

  const clientConfig = [{
    clientId: 'capitalone',
    apps: [{
      appId: 'chatbot',
      accountId: 'chatbot'
    }]
  }, {
    clientId: 'sirius',
    apps: [{
      appId: 'chatbot',
      accountId: 'chatbot'
    }]
  }];

  const unifiedPortalClientPicker = false;

  const {
    appReducer,
    types,
    stateKey
  } = appState;

  describe('App Actions:', function () {
    let store;

    beforeAll(() => {
      const middlewares = [thunk];
      const mockStore = configureStore(middlewares);
      store = mockStore(fromJS({
        app: {
          ...caa
        },
        projects: fromJS({
          projects: List([{ projectId: 'pro-123', projectName: 'project 123' }]),
          activeProjectId: 'pro-123',
          trackProjects: false,
        }),
        runs: fromJS({
          runs: List([{ runId: 'run-123', projectName: 'run 123' }]),
          selectedRunId: 'run-123',
          trackRuns: false,
        }),
        conversations: {
          selectedTranscriptId: ''
        }
      }));
    });

    afterEach(() => {
      jest.clearAllMocks();
      store.clearActions();
    });

    describe('getUserConfig:', () => {
      test('should dispatch getUserConfig', () => {
        axios.get.mockResolvedValue({
          data: mockClientConfigData
        });
        store.dispatch(appState.getUserConfig());
        expect(axios.get).toHaveBeenCalled();
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    describe('getConversationsConfig:', () => {
      test('should dispatch getConversationsConfig', () => {
        axios.get.mockResolvedValue({
          data: mockClientConfigData
        });
        store.dispatch(appState.getConversationsConfig());
        expect(axios.get).toHaveBeenCalled();
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    describe('getAnswersConfig:', () => {
      test('should dispatch getAnswersConfig', () => {
        axios.get.mockResolvedValue({
          data: mockClientConfigData
        });
        store.dispatch(appState.getAnswersConfig());
        expect(axios.get).toHaveBeenCalled();
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    describe('updateCAA:', () => {
      test('should dispatch update CAA', () => {
        store.dispatch(appState.updateCAA(caa));
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    describe('setUnifiedPortalClientPicker', () => {
      test('should dispatch setUnifiedPortalClientPicker action', () => {
        store.dispatch(appState.setUnifiedPortalClientPicker(unifiedPortalClientPicker));
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    describe('setUserConfig', () => {
      test('should dispatch setUserConfig action', () => {
        const userConfig = {
          authEnabled: false,
          analyticsKey: 'c8adb0c80da1f7ed082cfc43c4da0fe4',

        };
        store.dispatch(appState.setUserConfig(userConfig));
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    describe('setClientsConfig', () => {
      test('should dispatch setClientsConfig action', () => {
        store.dispatch(appState.setClientsConfig(clientConfig));
        expect(store.getActions()).toMatchSnapshot();
      });

      test('should dispatch setClientsConfig action', () => {
        const emptyClientConfig = [];
        store.dispatch(appState.setClientsConfig());
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    describe('setModalIsOpen', () => {
      test('should dispatch setModalIsOpen action', () => {
        store.dispatch(appState.setModalIsOpen(true, {
          modalName: Constants.modals.confirm,
          header: `${Constants.cancelHeader} project?`,
          message: `${Constants.cancelMessage} project?`,
        }));
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    describe('setModalIsOpen', () => {
      test('should dispatch setModalIsOpen action', () => {
        store.dispatch(appState.setModalIsOpen(true, {
          modalName: Constants.modals.confirm,
          header: `${Constants.cancelHeader} project?`,
          message: `${Constants.cancelMessage} project?`,
        }));
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    describe('setModalMessageState', () => {
      test('should dispatch setModalMessageState action', () => {
        store.dispatch(appState.setModalMessageState('Modal is Open'));
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    describe('setModalMessageState', () => {
      test('should dispatch setModalMessageState action', () => {
        store.dispatch(appState.setModalMessageState('Modal is Open'));
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    describe('setModalMessage', () => {
      test('should dispatch setModalMessage action', () => {
        store.dispatch(appState.setModalMessage('Modal is Open'));
        expect(store.getActions()).toMatchSnapshot();
      });
    });
  });

  describe('App Reducers: ', function () {

    test('should return the initial state', () => {
      const newState = appReducer(undefined, {
        type: types.NOOP,
      });
      expect(newState).toEqual(appState.initialAppState);
    });

    describe('SET_CLIENTS_CONFIG:', function () {
      let currentState;

      beforeAll(() => {
        currentState = fromJS({
          clientConfigs: [],
        });
      });

      test('should set the client config', () => {
        const newState = appReducer(currentState, {
          type: types.SET_CLIENTS_CONFIG,
          clientConfigs: clientConfig,
        });
        expect(newState.get(stateKey.clientConfigs).toJS()).toEqual(clientConfig);
      });
    });

    describe('SET_UNIFIED_PORTAL_CLIENT_PICKER:', function () {
      let currentState;

      beforeAll(() => {
        currentState = fromJS({
          unifiedPortalClientPicker: false,
        });
      });

      test('should set the UNIFIED_PORTAL_CLIENT_PICKER', () => {
        const newState = appReducer(currentState, {
          type: types.SET_UNIFIED_PORTAL_CLIENT_PICKER,
          unifiedPortalClientPicker: true,
        });
        expect(newState.get(stateKey.unifiedPortalClientPicker)).toEqual(true);
      });
    });

    describe('SET_USER_CONFIG:', function () {
      let currentState;
      let config = {
        userName: 'testup',
        userType: 'Mar@2020',
        authEnabled: true,
        itsBaseUrl: '',
        analyticsKey: '',
        docPortalUrl: '',
        unifiedPortalUrl: '',
        contactSupportUrl: '',
      }

      beforeAll(() => {
        currentState = fromJS({
          userName: '',
          userType: '',
          authEnabled: 'false',
          itsBaseUrl: '',
          analyticsKey: '',
          docPortalUrl: '',
          unifiedPortalUrl: '',
          contactSupportUrl: '',
        });
      });

      test('should set the SET_USER_CONFIG', () => {
        const newState = appReducer(currentState, {
          type: types.SET_USER_CONFIG,
          config,
        });
        expect(newState.get(stateKey.authEnabled)).toEqual(true);
      });
    });

    describe('MODAL_IS_OPEN:', function () {
      let currentState;
      let currentModalState = {
        modalName: ''
      };

      let modalState = {
        modalName: 'Modal Name'
      };

      beforeAll(() => {
        currentState = fromJS({
          modalIsOpen: false,
          modalState: fromJS(currentModalState),
        });
      });

      test('should set the MODAL_IS_OPEN', () => {
        const newState = appReducer(currentState, {
          type: types.MODAL_IS_OPEN,
          modalIsOpen: true,
          modalState: fromJS(modalState),
        });
        expect(newState.get(stateKey.modalIsOpen)).toEqual(true);
        expect(newState.get(stateKey.modalState).toJS()).toEqual(modalState);
      });
    });

    describe('MODAL_MESSAGE:', function () {
      let currentState;

      beforeAll(() => {
        currentState = fromJS({
          modalMessage: '',
        });
      });

      test('should set the MODAL_MESSAGE', () => {
        const newState = appReducer(currentState, {
          type: types.MODAL_MESSAGE,
          modalMessage: 'Modal Message',
        });
        expect(newState.get(stateKey.modalMessage)).toEqual('Modal Message');
      });
    });

    test('LOCATION_CHANGE', () => {
      const currentState = fromJS({
          clientId: '',
          appId: '',
          accountId: '',
          componentClientId: '',
          clientConfigs: List(),
      });
      const newState = appReducer(currentState, {
        type: GlobalTypes.LOCATION_CHANGE,
        payload: {
          action: 'POP',
          location: {
            search: '?clientid=testclientone&account=service&appid=general&componentClientId=testclientone'
          },
        }
      });
      expect(newState.get(stateKey.clientId)).toEqual('testclientone');
    });
  });

  describe('AppSelectors Selectors:', function () {
    let state;
    let caa = {
      clientId: 'tfsinc',
      appId: 'idt_test',
      accountId: 'service',
      componentClientId: '247inc',
    };
    let modalState = {
      modalName: 'Modal Name'
    }

    beforeAll(() => {
      state = fromJS({
        app: {
          authEnabled: false,
          clientId: 'tfsinc',
          componentClientId: '247inc',
          appId: 'idt_test',
          accountId: 'service',
          userName: 'testup',
          userType: 'INTERNAL',
          itsBaseUrl: 'https://stable.developer.sv2.247-inc.net/v1/integratedtoolsuite/',
          analyticsKey: '',
          clientConfigs: fromJS(clientConfig),
          modalIsOpen: false,
          modalMessage: 'Modal Message',
          modalState: fromJS(modalState),
          unifiedPortalClientPicker: false,
          docPortalUrl: 'https://stable.developer.sv2.247-inc.net/docportal/Content/Intent-Discovery/Intent-Discovery.htm',
          unifiedPortalUrl: 'https://stable.developer.sv2.247-inc.net/home/',
          contactSupportUrl: 'http://chainsaw.tellme.com/~achernous/cssp/scxml/#start',
          userRole: 'admin',
          roleConfig: {},
        }
      });      
    });

    test('getAppState', function () {
      expect(appState.getAppState(state)).toEqual(state.get('app'));
    });

    test('getCAASelector', function () {
      expect(appState.getCAASelector(state)).toEqual(caa);
    });

    test('getClientIdSelector', function () {
      expect(appState.getClientIdSelector(state)).toEqual('tfsinc');
    });

    test('getComponentClientIdSelector', function () {
      expect(appState.getComponentClientIdSelector(state)).toEqual('247inc');
    });

    test('getAppIdSelector', function () {
      expect(appState.getAppIdSelector(state)).toEqual('idt_test');
    });

    test('getAccountIdSelector', function () {
      expect(appState.getAccountIdSelector(state)).toEqual('service');
    });

    test('getClientConfigsSelector', function () {
      expect(appState.getClientConfigsSelector(state)).toEqual(fromJS(clientConfig));
    });

    test('getITSBaseUrlSelector', function () {
      expect(appState.getITSBaseUrlSelector(state)).toEqual('https://stable.developer.sv2.247-inc.net/v1/integratedtoolsuite/');
    });

    test('getModalOpenSelector', function () {
      expect(appState.getModalOpenSelector(state)).toEqual(false);
    });

    test('getModalStateSelector', function () {
      expect(appState.getModalStateSelector(state)).toEqual(fromJS(modalState));
    });

    test('getModalMessageSelector', function () {
      expect(appState.getModalMessageSelector(state)).toEqual('Modal Message');
    });

    test('getUnifiedPortalClientPicker', function () {
      expect(appState.getUnifiedPortalClientPicker(state)).toEqual(false);
    });

    test('getDocPortalUrlSelector', function () {
      expect(appState.getDocPortalUrlSelector(state)).toEqual('https://stable.developer.sv2.247-inc.net/docportal/Content/Intent-Discovery/Intent-Discovery.htm');
    });

    test('getUnifiedPortalUrlSelector', function () {
      expect(appState.getUnifiedPortalUrlSelector(state)).toEqual('https://stable.developer.sv2.247-inc.net/home/');
    });

    test('getUserTypeSelector', function () {
      expect(appState.getUserTypeSelector(state)).toEqual('INTERNAL');
    });

    test('getContactSupportUrlSelector', function () {
      expect(appState.getContactSupportUrlSelector(state)).toEqual('http://chainsaw.tellme.com/~achernous/cssp/scxml/#start');
    });

    test('getRoleConfigSelector', function () {
      expect(appState.getRoleConfigSelector(state)).toEqual(fromJS({}));
    });

    test('getUserRoleSelector', function () {
      expect(appState.getUserRoleSelector(state)).toEqual('admin');
    });
  });
});
