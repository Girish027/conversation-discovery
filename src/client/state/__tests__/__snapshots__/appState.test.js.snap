// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`state/App: App Actions: getAnswersConfig: should dispatch getAnswersConfig 1`] = `Array []`;

exports[`state/App: App Actions: getConversationsConfig: should dispatch getConversationsConfig 1`] = `Array []`;

exports[`state/App: App Actions: getUserConfig: should dispatch getUserConfig 1`] = `Array []`;

exports[`state/App: App Actions: setClientsConfig should dispatch setClientsConfig action 1`] = `
Array [
  Object {
    "clientConfigs": Array [
      Object {
        "apps": Array [
          Object {
            "accountId": "chatbot",
            "appId": "chatbot",
          },
        ],
        "clientId": "capitalone",
      },
      Object {
        "apps": Array [
          Object {
            "accountId": "chatbot",
            "appId": "chatbot",
          },
        ],
        "clientId": "sirius",
      },
    ],
    "type": "app/SET_CLIENTS_CONFIG",
  },
]
`;

exports[`state/App: App Actions: setClientsConfig should dispatch setClientsConfig action 2`] = `
Array [
  Object {
    "clientConfigs": Array [],
    "type": "app/SET_CLIENTS_CONFIG",
  },
]
`;

exports[`state/App: App Actions: setModalIsOpen should dispatch setModalIsOpen action 1`] = `
Array [
  Object {
    "modalIsOpen": true,
    "modalState": Object {
      "header": "undefined project?",
      "message": "undefined project?",
      "modalName": "ConfirmationDialog",
    },
    "type": "app/MODAL_IS_OPEN",
  },
]
`;

exports[`state/App: App Actions: setModalIsOpen should dispatch setModalIsOpen action 2`] = `
Array [
  Object {
    "modalIsOpen": true,
    "modalState": Object {
      "header": "undefined project?",
      "message": "undefined project?",
      "modalName": "ConfirmationDialog",
    },
    "type": "app/MODAL_IS_OPEN",
  },
]
`;

exports[`state/App: App Actions: setModalMessage should dispatch setModalMessage action 1`] = `
Array [
  Object {
    "modalMessage": "Modal is Open",
    "type": "app/MODAL_MESSAGE",
  },
]
`;

exports[`state/App: App Actions: setModalMessageState should dispatch setModalMessageState action 1`] = `
Array [
  Object {
    "modalMessage": "Modal is Open",
    "type": "app/MODAL_MESSAGE",
  },
]
`;

exports[`state/App: App Actions: setModalMessageState should dispatch setModalMessageState action 2`] = `
Array [
  Object {
    "modalMessage": "Modal is Open",
    "type": "app/MODAL_MESSAGE",
  },
]
`;

exports[`state/App: App Actions: setUnifiedPortalClientPicker should dispatch setUnifiedPortalClientPicker action 1`] = `
Array [
  Object {
    "type": "app/SET_UNIFIED_PORTAL_CLIENT_PICKER",
    "unifiedPortalClientPicker": false,
  },
]
`;

exports[`state/App: App Actions: setUserConfig should dispatch setUserConfig action 1`] = `
Array [
  Object {
    "config": Object {
      "analyticsKey": "c8adb0c80da1f7ed082cfc43c4da0fe4",
      "authEnabled": false,
    },
    "type": "app/SET_USER_CONFIG",
  },
]
`;

exports[`state/App: App Actions: updateCAA: should dispatch update CAA 1`] = `
Array [
  Object {
    "caa": Object {
      "accountId": "referencebot",
      "appId": "referencebot",
      "clientId": "247ai",
    },
    "type": "all/UPDATE_CAA",
  },
]
`;
